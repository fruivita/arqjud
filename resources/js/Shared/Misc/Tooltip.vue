<!--
    Exibe um tooltip ao se clicar no botÃ£o.

    @link https://vuejs.org/guide/introduction.html
    @link https://tailwindcss.com/docs
    @link https://inertiajs.com/
 -->

<script setup>
import Icone from '@/Shared/Icones/Icone.vue';
import { arrow, computePosition, flip, offset, shift } from '@floating-ui/dom';
import { onClickOutside } from '@vueuse/core';
import { isArray } from 'lodash';
import { ref } from 'vue';

const props = defineProps({
    icone: { type: String, default: 'info-circle' },
    texto: { type: [Array, String], required: true },
    posicao: {
        type: String,
        default: 'bottom',
        validator(value) {
            return ['bottom', 'top', 'left', 'right'].includes(value);
        },
    },
});

const referenceRef = ref(null);
const floatingRef = ref(null);
const arrowRef = ref(null);
const exibirTooltip = ref(false);
const clicked = ref(false);

const click = () => {
    exibirTooltip.value = true;
    clicked.value = true;
    calcularPosicao();
};

const mouveover = () => {
    if (exibirTooltip.value !== true) {
        exibirTooltip.value = true;
        calcularPosicao();
    }
};

const mouseout = () => {
    if (clicked.value !== true) {
        exibirTooltip.value = false;
    }
};

const calcularPosicao = async () => {
    const { x, y, middlewareData, placement } = await computePosition(
        referenceRef.value,
        floatingRef.value,
        {
            placement: props.posicao.value,
            middleware: [offset(8), flip(), shift(), arrow({ element: arrowRef.value })],
        }
    );

    Object.assign(floatingRef.value.style, {
        left: `${x}px`,
        top: `${y}px`,
    });

    const { x: arrowX, y: arrowY } = middlewareData.arrow;

    const ladoOposto = {
        left: 'right',
        right: 'left',
        bottom: 'top',
        top: 'bottom',
    }[placement.split('-')[0]];

    Object.assign(arrowRef.value.style, {
        left: arrowX ? `${arrowX}px` : '',
        top: arrowY ? `${arrowY}px` : '',
        bottom: '',
        right: '',
        [ladoOposto]: '-4px',
    });
};

onClickOutside(floatingRef, () => {
    exibirTooltip.value = false;
    clicked.value = false;
});
</script>

<template>
    <div class="inline-block">
        <button
            ref="referenceRef"
            @click="click"
            @mouseover="mouveover"
            @mouseout="mouseout"
            class="rounded p-1 outline-none ring-primaria-500 hover:ring-2 focus:ring-2 dark:ring-secundaria-500"
            dusk="toggle"
            type="button"
        >
            <Icone :nome="icone" class="h-4 w-4" />
        </button>

        <div
            v-show="exibirTooltip"
            ref="floatingRef"
            class="absolute top-0 left-0 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50"
            dusk="tooltip"
        >
            <template v-if="isArray(texto)">
                <span v-for="(txt, indice) in texto" :key="indice" class="block">{{ txt }}</span>
            </template>

            <template v-else>
                <span>{{ texto }}</span>
            </template>

            <div
                ref="arrowRef"
                class="absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600"
            ></div>
        </div>
    </div>
</template>
