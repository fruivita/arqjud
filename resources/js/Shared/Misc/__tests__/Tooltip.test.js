/**
 * Testes para o componente Tooltip.
 *
 * @link https://vuejs.org/guide/scaling-up/testing.html
 * @link https://test-utils.vuejs.org/guide/
 * @link https://vitest.dev/
 */

import Tooltip from '@/Shared/Misc/Tooltip.vue';
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';

let mountFunction;

beforeEach(() => {
    mountFunction = (options = {}) => {
        return mount(Tooltip, { ...options });
    };
});

describe('Tooltip', () => {
    // Caminho feliz
    const posicoes = ['bottom', 'top', 'left', 'right'];

    test('aceita apenas as posições definidas', () => {
        const validator = Tooltip.props.posicao.validator;

        posicoes.forEach((type) => expect(validator(type)).toBe(true));
        expect(validator('foo')).toBe(false);
    });

    test('propriedades do componente estão definidas', () => {
        expect(Tooltip.props).toMatchObject({
            icone: { type: String, default: 'info-circle' },
            texto: { type: [Array, String], required: true },
            posicao: { type: String, default: 'bottom' },
        });
    });

    test('renderiza o componente default respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: { texto: 'foo' },
            }).html()
        ).toMatchSnapshot();
    });

    test('renderiza o componente com ícone específico respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: { icone: 'key', texto: 'foo' },
            }).html()
        ).toMatchSnapshot();
    });

    test('renderiza o componente informando um array de textos respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: { texto: ['foo', 'bar', 'baz'] },
            }).html()
        ).toMatchSnapshot();
    });

    test.each(posicoes)(
        'renderiza o componente nas posições específicas respeitando o snapshot',
        (posicao) => {
            expect(
                mountFunction({
                    props: { texto: 'foo', posicao: posicao },
                }).html()
            ).toMatchSnapshot();
        }
    );

    test('exibe o tooltip ao clicar no botão respeitando o snapshot', async () => {
        const wrapper = mountFunction({
            props: { texto: 'foo' },
        });

        await wrapper.find('[dusk="toggle"]').trigger('click');

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('exibe o tooltip ao clicar no botão', async () => {
        const wrapper = mountFunction({
            props: { texto: 'foo' },
        });

        const tooltip = wrapper.get('[dusk="tooltip"]');

        expect(tooltip.isVisible()).toBeFalsy();

        await wrapper.get('[dusk="toggle"]').trigger('click');

        expect(tooltip.isVisible()).toBeTruthy();
    });

    test('alterna a visibilidade do tooltip nos ventos mouseover e mouseout no botão', async () => {
        const wrapper = mountFunction({
            props: { texto: 'foo' },
        });

        const tooltip = wrapper.get('[dusk="tooltip"]');

        expect(tooltip.isVisible()).toBeFalsy();

        await wrapper.get('[dusk="toggle"]').trigger('mouseover');

        expect(tooltip.isVisible()).toBeTruthy();

        await wrapper.get('[dusk="toggle"]').trigger('mouseout');

        expect(tooltip.isVisible()).toBeFalsy();
    });

    test('mouseover e mouseout não possuem efeito se houver um click para exibição', async () => {
        const wrapper = mountFunction({
            props: { texto: 'foo' },
        });

        const tooltip = wrapper.get('[dusk="tooltip"]');

        expect(tooltip.isVisible()).toBeFalsy();

        await wrapper.get('[dusk="toggle"]').trigger('click');

        expect(tooltip.isVisible()).toBeTruthy();

        await wrapper.get('[dusk="toggle"]').trigger('mouseover');

        expect(tooltip.isVisible()).toBeTruthy();

        await wrapper.get('[dusk="toggle"]').trigger('mouseout');

        expect(tooltip.isVisible()).toBeTruthy();
    });
});
