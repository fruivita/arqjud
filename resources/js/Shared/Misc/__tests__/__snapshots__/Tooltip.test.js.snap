// Vitest Snapshot v1

exports[`Tooltip > exibe o tooltip ao clicar no botão respeitando o snapshot 1`] = `
"<div class=\\"inline-block\\"><button class=\\"rounded p-1 outline-none ring-primaria-500 hover:ring-2 focus:ring-2 dark:ring-secundaria-500\\" dusk=\\"toggle\\" type=\\"button\\"><svg class=\\"h-5 w-5\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path d=\\"M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z\\"></path>
      <path d=\\"m8.93 6.588-2.29.287-.082.38.45.083c.294.07.352.176.288.469l-.738 3.468c-.194.897.105 1.319.808 1.319.545 0 1.178-.252 1.465-.598l.088-.416c-.2.176-.492.246-.686.246-.275 0-.375-.193-.304-.533L8.93 6.588zM9 4.5a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
    </svg></button>
  <div class=\\"absolute top-0 left-0 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"tooltip\\" style=\\"\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Tooltip > renderiza o componente com ícone específico respeitando o snapshot 1`] = `
"<div class=\\"inline-block\\"><button class=\\"rounded p-1 outline-none ring-primaria-500 hover:ring-2 focus:ring-2 dark:ring-secundaria-500\\" dusk=\\"toggle\\" type=\\"button\\"><svg class=\\"h-5 w-5\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path d=\\"M0 8a4 4 0 0 1 7.465-2H14a.5.5 0 0 1 .354.146l1.5 1.5a.5.5 0 0 1 0 .708l-1.5 1.5a.5.5 0 0 1-.708 0L13 9.207l-.646.647a.5.5 0 0 1-.708 0L11 9.207l-.646.647a.5.5 0 0 1-.708 0L9 9.207l-.646.647A.5.5 0 0 1 8 10h-.535A4 4 0 0 1 0 8zm4-3a3 3 0 1 0 2.712 4.285A.5.5 0 0 1 7.163 9h.63l.853-.854a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.646-.647a.5.5 0 0 1 .708 0l.646.647.793-.793-1-1h-6.63a.5.5 0 0 1-.451-.285A3 3 0 0 0 4 5z\\"></path>
      <path d=\\"M4 8a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
    </svg></button>
  <div class=\\"absolute top-0 left-0 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"tooltip\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Tooltip > renderiza o componente default respeitando o snapshot 1`] = `
"<div class=\\"inline-block\\"><button class=\\"rounded p-1 outline-none ring-primaria-500 hover:ring-2 focus:ring-2 dark:ring-secundaria-500\\" dusk=\\"toggle\\" type=\\"button\\"><svg class=\\"h-5 w-5\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path d=\\"M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z\\"></path>
      <path d=\\"m8.93 6.588-2.29.287-.082.38.45.083c.294.07.352.176.288.469l-.738 3.468c-.194.897.105 1.319.808 1.319.545 0 1.178-.252 1.465-.598l.088-.416c-.2.176-.492.246-.686.246-.275 0-.375-.193-.304-.533L8.93 6.588zM9 4.5a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
    </svg></button>
  <div class=\\"absolute top-0 left-0 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"tooltip\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Tooltip > renderiza o componente informando um array de textos respeitando o snapshot 1`] = `
"<div class=\\"inline-block\\"><button class=\\"rounded p-1 outline-none ring-primaria-500 hover:ring-2 focus:ring-2 dark:ring-secundaria-500\\" dusk=\\"toggle\\" type=\\"button\\"><svg class=\\"h-5 w-5\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path d=\\"M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z\\"></path>
      <path d=\\"m8.93 6.588-2.29.287-.082.38.45.083c.294.07.352.176.288.469l-.738 3.468c-.194.897.105 1.319.808 1.319.545 0 1.178-.252 1.465-.598l.088-.416c-.2.176-.492.246-.686.246-.275 0-.375-.193-.304-.533L8.93 6.588zM9 4.5a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
    </svg></button>
  <div class=\\"absolute top-0 left-0 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"tooltip\\" style=\\"display: none;\\"><span class=\\"block\\">foo</span><span class=\\"block\\">bar</span><span class=\\"block\\">baz</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Tooltip > renderiza o componente nas posições específicas respeitando o snapshot 1`] = `
"<div class=\\"inline-block\\"><button class=\\"rounded p-1 outline-none ring-primaria-500 hover:ring-2 focus:ring-2 dark:ring-secundaria-500\\" dusk=\\"toggle\\" type=\\"button\\"><svg class=\\"h-5 w-5\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path d=\\"M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z\\"></path>
      <path d=\\"m8.93 6.588-2.29.287-.082.38.45.083c.294.07.352.176.288.469l-.738 3.468c-.194.897.105 1.319.808 1.319.545 0 1.178-.252 1.465-.598l.088-.416c-.2.176-.492.246-.686.246-.275 0-.375-.193-.304-.533L8.93 6.588zM9 4.5a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
    </svg></button>
  <div class=\\"absolute top-0 left-0 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"tooltip\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Tooltip > renderiza o componente nas posições específicas respeitando o snapshot 2`] = `
"<div class=\\"inline-block\\"><button class=\\"rounded p-1 outline-none ring-primaria-500 hover:ring-2 focus:ring-2 dark:ring-secundaria-500\\" dusk=\\"toggle\\" type=\\"button\\"><svg class=\\"h-5 w-5\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path d=\\"M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z\\"></path>
      <path d=\\"m8.93 6.588-2.29.287-.082.38.45.083c.294.07.352.176.288.469l-.738 3.468c-.194.897.105 1.319.808 1.319.545 0 1.178-.252 1.465-.598l.088-.416c-.2.176-.492.246-.686.246-.275 0-.375-.193-.304-.533L8.93 6.588zM9 4.5a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
    </svg></button>
  <div class=\\"absolute top-0 left-0 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"tooltip\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Tooltip > renderiza o componente nas posições específicas respeitando o snapshot 3`] = `
"<div class=\\"inline-block\\"><button class=\\"rounded p-1 outline-none ring-primaria-500 hover:ring-2 focus:ring-2 dark:ring-secundaria-500\\" dusk=\\"toggle\\" type=\\"button\\"><svg class=\\"h-5 w-5\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path d=\\"M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z\\"></path>
      <path d=\\"m8.93 6.588-2.29.287-.082.38.45.083c.294.07.352.176.288.469l-.738 3.468c-.194.897.105 1.319.808 1.319.545 0 1.178-.252 1.465-.598l.088-.416c-.2.176-.492.246-.686.246-.275 0-.375-.193-.304-.533L8.93 6.588zM9 4.5a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
    </svg></button>
  <div class=\\"absolute top-0 left-0 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"tooltip\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Tooltip > renderiza o componente nas posições específicas respeitando o snapshot 4`] = `
"<div class=\\"inline-block\\"><button class=\\"rounded p-1 outline-none ring-primaria-500 hover:ring-2 focus:ring-2 dark:ring-secundaria-500\\" dusk=\\"toggle\\" type=\\"button\\"><svg class=\\"h-5 w-5\\" fill=\\"currentColor\\" viewBox=\\"0 0 16 16\\" xmlns=\\"http://www.w3.org/2000/svg\\">
      <path d=\\"M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z\\"></path>
      <path d=\\"m8.93 6.588-2.29.287-.082.38.45.083c.294.07.352.176.288.469l-.738 3.468c-.194.897.105 1.319.808 1.319.545 0 1.178-.252 1.465-.598l.088-.416c-.2.176-.492.246-.686.246-.275 0-.375-.193-.304-.533L8.93 6.588zM9 4.5a1 1 0 1 1-2 0 1 1 0 0 1 2 0z\\"></path>
    </svg></button>
  <div class=\\"absolute top-0 left-0 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"tooltip\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;
